[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict.

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77628

As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

fegin committed:
[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict.

As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_optim_state.py:
```diff
@@ -504,43 +539,71 @@ def test_scatter_full_optim_state_dict_nested(
-    def _test_shard_full_optim_state(
+    @skip_if_lt_x_gpu(2)
+    @parametrize("use_multiple_param_groups", [False, True])
+    def test_flatten_sharded_optim_state_dict_nested(
```
Would I be correct to say that these added tests do not exercise resharding to different world size? They are just sanity checking that you can save and load a fixed sharded optim state dict under the same process group configuration?

fegin replied:
Resharding sharded_optim_state is not supported yet. So yes, these tests do not test resharding case. These tests test the optimizer values and keys are correct. It is redundant for `test_flatten_sharded_optim_state_dict_transformer()` to create two models as they are the same.  But `test_flatten_sharded_optim_state_dict_nested()` needs to create two models to test wrap alternative feature. Reusing `_test_load_optim_state` to avoid duplicating the test code.

awgu commented on test/distributed/fsdp/test_fsdp_optim_state.py:
```diff
@@ -628,8 +700,9 @@ def test_shard_full_optim_state_dict_unmanaged_params(
-        into the same subroutine :meth:`_flatten_full_optim_state_dict`.
+        :meth:`scatter_full_optim_state_dict` and `flatten_sharded_optim_state_dict`
+        to save CI cost since it call into the same subroutine
+        :meth:`_flatten_full_optim_state_dict`.
```
```suggestion
        :meth:`_flatten_optim_state_dict`.
```

(since you renamed it)

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -173,6 +188,7 @@ def _unflatten_communicated_optim_state(
     fsdp_module,
     flat_param: FlatParameter,
     state: _ConsolidatedOptimState,
+    shard_state,
```
nit:
```suggestion
    shard_state: bool,
```

awgu reviewed (commented):
The overall logic looks good to me. I just left a few minor comments. I will wait for others to take a look before approving.

fegin commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1000,3 +1026,142 @@ def _get_unflat_to_flat_param_ids(
+    The flattened parameters in ``FSDP`` modules contained in ``model``
+    are mapped back to their unflattened parameters.
+
+    .. warning:: This needs to be called on all ranks since synchronization
```
These warnings are still in the public API. I'll removed the warnings for this private API.

fegin commented on test/distributed/fsdp/test_fsdp_optim_state.py:
```diff
@@ -36,6 +43,13 @@ class _OSDCommMethod(Enum):
+    FLATTEN_SHARDED_OSD = auto()
+
+
+class _ModelClass(Enum):
```
This enum is not used by parametrize(). I still kept the original individual tests to test different model classes. This enum is only used to pass the model class to the internal test API, `_test_load_optim_state()`. Creating this enum to replace str.

awgu commented:
Thanks for addressing all of my feedback so quickly! I will give @zhaojuanmao a chance to review, but the PR looks good to me.

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3482,83 +3476,50 @@ def full_optim_state_dict(
+        """
+        The API is similar to :meth:``full_optim_state_dict`` but chunks all
+        non-zero-dimension states to ShardedTensor to save memory. This API
+        should only be used when the state_dict is sharded_state_dict.
```
How does the API break when state_dict is not sharded state dict? Do we check this anywhere / report appropriate errors?

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3621,8 +3582,48 @@ def shard_full_optim_state_dict(
+    ) -> Dict[str, Any]:
+        """
+        The API is similar to :meth:``shard_full_optim_state_dict``. The only
+        difference is that the input ``full_optim_state_dict`` should be returned
```
Not sure if I'm missing something, but the doc might not be correct? i.e. the input should be from `sharded_optim_state_dict`? let me know if I'm missing something

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3482,83 +3476,50 @@ def full_optim_state_dict(
+        doc.
+
+        .. warning:: The returned state dict contains ShardedTensor and cannot be
+            directly used by the regular ``optim.load_state_dict``.
```
Are there docs on how to load an optim state_dict obtained via `sharded_optim_state_dict`?

fegin replied:
The usage is exact the same as full_optim_state_dict. So the docstring just refers that of `full_optim_state_dict`.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3482,83 +3476,50 @@ def full_optim_state_dict(
+        # and one API to load optimizer states. ``state_dict_type`` will be used
+        # to decide which optimizer states should be returned.
+        # There are currently two APIs to load a full optimizer state. So the
+        # first step of the unification is to merge the two full optimizer state
```
cc @awgu for any thoughts regarding consolidation. In general, I believe we should do this and offer clear docs on how to use such APIs, and instruct users as to how to use model state_dict and optim_state_dict (and their sharded counterparts) appropriately together.

rohan-varma commented on torch/distributed/fsdp/shard_utils.py:
```diff
@@ -150,16 +154,39 @@ def _all_gather_sharded_tensor(
+    dist._all_gather_base(all_move_to_cpu, move_to_cpu, group=pg)
+
+    tensor = tensor.narrow(0, 0, tensor_numel).reshape(sharded_tensor.size())
+    if all_move_to_cpu.sum().is_nonzero():
```
Curious, is it possible that `local_tensor` will be on CPU / not on CPU for some ranks and not others, and if so how is this possible?

fegin replied:
I actually cannot think one that some ranks are on CPU but others are on GPU. I did this for precaution to avoid some corner cases. But may be it is better just error out if not this case happens?

rohan-varma commented on torch/distributed/fsdp/shard_utils.py:
```diff
@@ -174,17 +201,26 @@ def _gather_state_dict(
     return new_state_dict
+
+
+def _distributed_chunk_tensor(
```
If this is generally useful, can we see if this can be offered natively by the ShardedTensor package?

fegin replied:
Discussing with @wanchaol , will upstream some APIs in shard_utils if they are generic for other use cases.

rohan-varma reviewed (commented):
Looks good overall! Mainly commented around `shard_state` changes as that seems to be the crux of the change, and it seems that some of the full_osd code was just moved over. Thank you!

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3622,8 +3583,48 @@ def shard_full_optim_state_dict(
+            Refer to :meth:``shard_full_optim_state_dict``.
+        """
+
+        # TODO: The implementation is the same as ``shard_full_optim_state_dict``.
+        # See the TODO in ``shard_full_optim_state_dict`` for the future
+        # unification plan.
```
let's file a github issue? right now I feel the API names overall are a little confusing

zhaojuanmao commented on torch/distributed/fsdp/_shard_utils.py:
```diff
@@ -150,16 +154,39 @@ def _all_gather_sharded_tensor(
+
+    tensor = tensor.narrow(0, 0, tensor_numel).reshape(sharded_tensor.size())
+    if all_move_to_cpu.sum().is_nonzero():
+        tensor = tensor.cpu()
```
just realized that maybe we should not move the all gathered data to cpu here, otherwise all ranks are moving redundant all gathered data to cpu and could make cpu OOM. instead, we may be able to move the data back to cpu after sharding as needed.

fegin replied:
Removed the "moving back to CPU" part and let users to handle it.

zhaojuanmao reviewed (commented):
left some minor comments

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3723,115 +3718,53 @@ def full_optim_state_dict(
+        """
+        The API is similar to :meth:``full_optim_state_dict`` but this API
+        chunks all non-zero-dimension states to ShardedTensor to save memory.
+        This API should only be used when the state_dict is sharded_state_dict.
```
Maybe this line is mis-stated? If I am not mistaken, this is `sharded_optim_state_dict()`, which does not have a `state_dict` argument.

fegin replied:
The description is indeed misleading. I changed the description in the latest verion.

awgu reviewed (approved):
LGTM!

I agree with everyone's thoughts throughout this PR about the confusion around optimizer state dict APIs/docs/usage. At the moment, it is not clear to me what the future for optimizer state checkpointing (and even model state checkpointing) for FSDP should look like. I would love to sync about it with everyone at some point.

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge

fegin committed:
Update on "[FSDP] Implement sharded_optim_state_dict and flatten_sharded_optim_state_dict."


As title

Differential Revision: [D36436496](https://our.internmc.facebook.com/intern/diff/D36436496/)

[ghstack-poisoned]
